### ISSUE TITLE
Instanced CollisionPolygon2D's cannot perform collision detection

### ISSUE THREAD

**Issue by JarLowrey** [godotengine/godot#8896]
**Operating system or device - Godot version:**

Ubuntu 16.06, Windows 8.1, Godot 2.1.3

**Issue description:**

Save a collision polygon as a scene, instance it thru scripting onto a rigidbody2d. Enable visible collision shapes. The shapes will appear correct, but collision checking fails. [QA context](https://godotengine.org/qa/15197/rigidbody2ds-not-colliding?state=comment-15197&show=15197#c15197)

**Link to minimal example project:**

GitHub upload is not working but you can view the project [here](https://ufile.io/1ojmq)

**Comment by JarLowrey** [comment]
This appears to be an issues with my polygons. When I added it directly as a node instead of instancing via scripting it still did not work. I recreated it using the editor and suddenly it was working. I would like to save the working CollisionPolygon2D to compare the xml with the broken one, but the Resource's save icon is grayed out. Is there any way to fix that?

**Comment by JarLowrey** [comment]
Broken polygon: 
```

<?xml version="1.0" encoding="UTF-8" ?>
<resource_file type="PackedScene" subresource_count="1" version="2.1" version_name="Godot Engine v2.1.3.stable.official">
	<main_resource>
		<dictionary name="_bundled" shared="false">
			<string> "conn_count" </string>
			<int> 0 </int>
			<string> "conns" </string>
			<int_array  len="0"> 				 </int_array>
			<string> "editable_instances" </string>
			<array  len="0" shared="false">
			</array>
			<string> "names" </string>
			<string_array  len="5">
				<string> "CollisionPolygon2D" </string>
				<string> "build_mode" </string>
				<string> "polygon" </string>
				<string> "shape_range" </string>
				<string> "trigger" </string>
			</string_array>
			<string> "node_count" </string>
			<int> 1 </int>
			<string> "node_paths" </string>
			<array  len="0" shared="false">
			</array>
			<string> "nodes" </string>
			<int_array  len="15"> 				-1, -1, 0, 0, -1, 4, 1, 0, 2, 1, 3, 2, 4, 3, 0 </int_array>
			<string> "variants" </string>
			<array  len="4" shared="false">
				<int> 0 </int>
				<vector2_array  len="11"> 					16.5, 16, -17.5, 16, -10.5, 37, -25.5, 42, -47.5, -16, -29.5, -42, -1.5, -29, 29.5, -42, 46.5, -17, 24.5, 44, 9.5, 36 </vector2_array>
				<vector2> -1, -1 </vector2>
				<bool> False </bool>
			</array>
			<string> "version" </string>
			<int> 2 </int>
		</dictionary>

	</main_resource>
</resource_file>

```
Working polygon: 
```

<?xml version="1.0" encoding="UTF-8" ?>
<resource_file type="PackedScene" subresource_count="1" version="2.1" version_name="Godot Engine v2.1.3.stable.official">
	<main_resource>
		<dictionary name="_bundled" shared="false">
			<string> "conn_count" </string>
			<int> 0 </int>
			<string> "conns" </string>
			<int_array  len="0"> 				 </int_array>
			<string> "editable_instances" </string>
			<array  len="0" shared="false">
			</array>
			<string> "names" </string>
			<string_array  len="5">
				<string> "CollisionPolygon2D" </string>
				<string> "build_mode" </string>
				<string> "polygon" </string>
				<string> "shape_range" </string>
				<string> "trigger" </string>
			</string_array>
			<string> "node_count" </string>
			<int> 1 </int>
			<string> "node_paths" </string>
			<array  len="0" shared="false">
			</array>
			<string> "nodes" </string>
			<int_array  len="15"> 				-1, -1, 0, 0, -1, 4, 1, 0, 2, 1, 3, 2, 4, 3, 0 </int_array>
			<string> "variants" </string>
			<array  len="4" shared="false">
				<int> 0 </int>
				<vector2_array  len="12"> 					-28.9916, -41.6547, -11.8205, -33.713, 10.0727, -33.4984, 27.6732, -41.6547, 46.5614, -16.1126, 25.0975, 41.4108, 10.7167, 36.4741, 15.6534, 16.298, -16.9718, 16.0833, -10.5326, 36.2594, -25.9867, 42.2693, -47.8799, -16.5419 </vector2_array>
				<vector2> -1, -1 </vector2>
				<bool> False </bool>
			</array>
			<string> "version" </string>
			<int> 2 </int>
		</dictionary>

	</main_resource>
</resource_file>
```

[updated project](https://ufile.io/1eyhw)

According to diffchecker.com the only difference btw the two files is in the vector2_array, which appears fine to me but idk what I'm looking at really

**Comment by JarLowrey** [comment]
Changing up the collision polygons somehow fixed it in my little test project, but I couldnt replicate in my actual project. I started another [(simpler) test project](https://ufile.io/99awv) and cannot get it to work. And the [working test project](https://ufile.io/2mpgf).

**Comment by JarLowrey** [comment]
Ok I think I've figured out how it's working in one project and not the other, but IDK why.

Create a new scene with a CollisionPolygon2D as root, draw something, save it as `polygon.tscn`. 

Create a new scene, make root RigidBody2D and save as `body.tscn`. Add a Sprite child and load its texture. Attach a script to Sprite, and in _ready call `get_parent().call_deferred("add_child",load("res://polygon.tscn").instance())`. 

Create a main scene and load multiple bodies. Set "Visible Collision Shapes" to ON. Run. Shapes are visible, but collisions don't work. 

Go back to `body.tscn`. Add a CollisionPolygon2D, then delete it. Run scene again, collisions work.

**Comment by eon-s** [comment]
On the example project you are trying to load an xml as a resource, Godot does not know what is that xml.
And in the case you create a valid resource, I do not see any attempt to set the loaded polygon as a ConcaveShape2D on the body either (loading on the helper only and not on the body is a common mistake).

**Comment by JarLowrey** [comment]
Yeah I am saving the CollisionPolygon2D as a scene and trying to instance it directly under a RigidBody2D. Should I instead copy over the CP2D's polygon to a Polygon2D or something?

**Comment by eon-s** [comment]
@JarLowrey well, you can't just copy the polygon, I think CollisonPolygon2D does not have a simple way to retrieve the shape, you may need to create a ConcaveShape2D, append the segments, then add it to the body.

**Comment by JarLowrey** [comment]
Currently testing the below code with a convex shaped CollisionPolyon2D scene, same issue - visible polygons but no collision. This time an error is thrown - " Index p_shape_idx out of size (shapes.size())." when I attempt to `	get_node("../CollisionShape2D").set_shape(shape)`.

```

func _ready():
	var collider = load("res://polygon.tscn").instance()
	var polygon = collider.get_polygon()
	var shape = null
	
	if collider.get_build_mode() == 0:
		shape = ConvexPolygonShape2D.new()
		shape.set_points(polygon)
	else:
		shape = ConcavePolygonShape2D.new()
		shape.set_segments(polygon) #causes issues
	
	get_node("../CollisionShape2D").set_shape(shape)
	pass
```

**Comment by puppetmaster-** [comment]
This is how I add a CollisionPoligon2D in my TileMapPlugin
`func setCollisionPolygon(seletedNode):
	if seletedNode.get_owner() != null:
		var _owner = seletedNode.get_owner()
		var _newStaticBodyNode = setStaticBody(seletedNode,_owner)
		var _newCollisionPolygonNode = CollisionPolygon2D.new()
		_newCollisionPolygonNode.set_polygon(getVector2ArrayFromSprite(seletedNode))
		_newCollisionPolygonNode.set_name("CollisionPolygon2D")
		_newCollisionPolygonNode.set_meta("_edit_lock_",true)
		_newStaticBodyNode.add_child(_newCollisionPolygonNode)
		_newStaticBodyNode.set_meta("_edit_lock_",true)
		_newCollisionPolygonNode.set_owner(_owner)`

https://godotengine.org/asset-library/asset/9

**Comment by JarLowrey** [comment]
This seems pretty similar to what I'm doing, IDK why it's failing for me.

**Comment by puppetmaster-** [comment]
do you set owner?

**Comment by eon-s** [comment]
I was able to make a concave shape added by script on a rigid body to collide with other shapes but not with another concave shape added by script (maybe I'm missing something).

**Comment by JarLowrey** [comment]
I did not set owner originally, but a [few comments ago](https://github.com/godotengine/godot/issues/8896#issuecomment-303832958) I am trying to extract the polygon from the CollisionPolygon2D and apply it to a CollisionShape2D that is added via the editor. That wasn't working either.

**Comment by eon-s** [comment]
Ok, made a couple of concave shapes with the editor (build mode on segment) and they ignore each other but collide with the rest of the shapes, is not just a script thing.
But this may be a problem related to other issues with concave shapes (I know this is going to change on 3).

The main issue is the way CollisionPolygon2D is being used outside the editor, you may better get rid of that node and work only on the bodies.

**Comment by JarLowrey** [comment]
Couldn't get anything to work - work around:

Create a RigidBody2D node with sprite and CollisionPolygon2D children. Instance the body where needed instead of the collision polygon. Switch out the bodies by calling this function:

```

func _change_body(body_path,script_path):
	var old_body = get_node("RigidBody2D")
	old_body.set_name("delete_me_thx_Aaaaaaaaa")
	var new_body = load(body_path).instance()
	new_body.set_name("RigidBody2D")
	_init_body(new_body)
	
	if old_body.has_node("Camera2D"):
		var cam = old_body.get_node("Camera2D")
		old_body.remove_child(cam)
		new_body.add_child(cam)
	
	add_child(new_body)
	if script_path and script_path.length() > 0:
		new_body.set_script(load(script_path))
	old_body.free()
```

**Comment by puppetmaster-** [comment]
If you only want to change a collisionShape durring runtime you can use this:

![image](https://cloud.githubusercontent.com/assets/10463138/26462364/c0380ae8-4180-11e7-9a23-e16973114b31.png)
```
func _ready():
	shapetransform = get_shape_transform(0)
	shape_ball = get_shape(0)
	shape_pad = get_shape(1)
	shape_tri = get_shape(2)
	shape_square = get_shape(3)
	mode = "normal"
	set_shape(mode)

func set_shape(newMode):
	hide_all_shapes()
	if (newMode == "normal"):
		get_node("shapes/normal").show()
		add_shape(shape_ball,shapetransform)
	elif(newMode == "pad"):
		get_node("shapes/pad").show()
		add_shape(shape_pad,shapetransform)
	elif(newMode == "tri"):
		get_node("shapes/tri").show()
		add_shape(shape_tri,shapetransform)
	elif(newMode == "square"):
		get_node("shapes/square").show()
		add_shape(shape_square,shapetransform)
	else:
		get_node("shapes/normal").show()
		add_shape(shape_ball,shapetransform)
	mode = newMode

func hide_all_shapes():
	clear_shapes()
	get_node("shapes/normal").hide()
	get_node("shapes/pad").hide()
	get_node("shapes/tri").hide()
	get_node("shapes/square").hide()
```

**Comment by kubecz3k** [comment]
I think this should be fixed now. Can you recheck?

**Comment by zejji** [comment]
I am also running into this issue with Godot 2.1.3.

Basically, I am trying to use the same collision shapes for different purposes. I am making a game which involves positioning of blocks. When the player is in the process of positioning the blocks, the collision shape needs to be used as a child of an Area2D node (i.e. to detect overlap but not affect physics) and, once positioned, it needs to be used as a child of a StaticBody2D (since the block then forms a part of the game world). The obvious way to do this would be by instancing a collision shape, but as noted above, this does not work. Without this functionality, it does not seem to be possible to reuse collision shapes, unless I am missing something...

@kubecz3k - When you say this should be fixed now, do you mean only in Godot 3.0? Do you know if there is there any plan to fix this in Godot 2.1.4 as this is literally a game-breaking bug for me :(

**Comment by KScorp** [comment]
I have the opposite issue, which I'm assuming is closely tied to this one. If I instance an Area2D scene, and add a CollisionShape2D as a child to that instance, the Area2D scene doesn't have any collision detection. If I discard instancing on the Area2D scene then collision works fine.

**Comment by RandomShaper** [comment]
It's been a while, but please test #15607. If my guess is right, it should fix this kind of issues.

**Comment by akien-mga** [comment]
It's the same issue as #13835 fixed in master by #15606. The 2.1 version should be fixed by #15607.
